### ISSUE TITLE
iup-rust and kiss-ui

### ISSUE THREAD

**Issue by thelink2012** [cybergeek94/kiss-ui#21]
**TL;DR:** Can and **should** kiss-ui and iup-rust coexist i.e. they have different goals?

I have been working on [overhauling iup-rust][1] this week to be idiomatic Rust instead of just a low-level wrappers, in the middle of this task I ended up finding this library (very nice btw) which builds upon IUP, instead of dropping all my work I choose to continue and see what happens.

But it still bites my mind that there's two kinda of similar libraries (at least at this moment that they are young), so I really needed to discuss this out over here to clear up my mind. I'm too new to the open source community and I'm not sure how everything ties together.

What's the direction this library is taking? Binding IUP or building a completly independent widget toolkit that just has IUP as backend? If not the later what should be done about two libraries doing the exactly *same* thing?

Best regards, nice library you are building here in a very quick pace :)

[1]: https://github.com/dcampbell24/iup-rust/pull/10

**Comment by cybergeek94** [comment]
I actually owe `iup-rust` some credit, as I copied the `iup-sys` bindings from there so I didn't have to write or generate my own. I had to comment out some stuff because the bindings weren't compiling at the time I started the project. If the author ever publishes them on Crates.io I'll be sure to use those instead.

When I started KISS-UI, `iup-rust` looked like it hadn't been updated in a while and still had a lot more work to be done. I thought it would be better to start working on my own library atop IUP than hope @dcampbell24 would come back and update it/merge PRs. I see he's back working on it now, which is really cool. I'd help out if I didn't have my own library to work on now.

I think KISS-UI and `iup-rust` are going in pretty different directions. I'm not only binding IUP, I'm also trying to give it a more Rust-y interface, which means abstracting over a lot of IUP's idioms and throwing some others out for the sake of the API. For example, using closures as callbacks and adding some glue code so they work with IUP, or abstracting over the string-y properties of IUP classes with setters.

If `iup-rust` continues in its current direction as a direct 1:1 Rust binding of IUP, then we won't have much collision. Hell, if `iup-rust` gets to the point where it covers a majority of IUP's API, then I could build KISS-UI on top of `iup-rust` directly. 

That's one great thing about open-source, is that competition and collaboration can exist concurrently. `iup-rust` and KISS-UI can work cooperatively while being technically competing solutions. We can share code and ideas, or straight-up copy them from one-another. I'm fine with you guys taking whatever you need from KISS-UI for `iup-rust`, it's only fitting since I took your FFI bindings.

But I'm also considering a future iteration that directly calls into native APIs instead of relying on IUP to abstract out the details. There's still the issue that IUP lacks a mature native driver for OS X, which is a large demographic in Rust. Sure, GTK+ works, but I was told on Reddit that it doesn't integrate with OS X well and "it looks like arse". So one day, KISS-UI might not even be related to `iup-rust` anymore, except in legacy.

**Comment by thelink2012** [comment]
I see, so I guess they should coexist with mutualism.

About iup-rust taking the 1:1 path, it's not anymore as of this overhaul. Now it uses traits and structs for elements representation, and yep it also accepting Rust closures as callbacks. (see the [counter](https://github.com/dcampbell24/iup-rust/blob/overhaul/examples/counter.rs) example for instance). I guess there are plans for strongly typed setters at some point but for now I think they are not top priority as the string setters work just fine and there are lots of things to do already. I assume the way it's going now is doing the bindings as Rusty as possible but without removing he IUP nomenclatures and feel in a way that you'd be able to code on iup-rust by looking in the C documentation just fine.

Cooperation of ideas already happened. You see, I had this problem of wanting a way to [build uis expressively in the code](https://github.com/dcampbell24/iup-rust/issues/7) _(note pseudo code in this issue is before the overhaul)_, your library does this quite well with the `children!` and returning ownership of objects in setters, and that is definitely something I inspired from in the overhaul even though I tried to look as little as possible into kiss-ui so that the directions they go don't _conflict_.

I would by the way suggest you to take a look at the callback system in iup-rust. It seems more flexible to add new callbacks than currently is in kiss-ui, so I'd suggest you to borrow the ideia or prehaps take it directly as it's "plugin 'n play" eh?

Your documentation is going pretty well, particularly the readmes that explains very very well how to install on each platform, you should really do a PR to add that in iup-rust readme. Oh, by the way, static linking is not really possible with the current Rust toolchain as it needs to compile a `.rc` file and link to it too (not only the static `.a`), meanwhile the `.dll` version has this `.rc` built in. This rc file contains the layout for some dialogs and other minor things.

**Comment by cybergeek94** [comment]
As far as callbacks go, it looks like we're doing pretty much the same thing. Adding new callbacks isn't really difficult. Since most IUP callback functions just take the `*mut Ihandle`, it's easy to abstract out most of the details: https://github.com/cybergeek94/kiss-ui/blob/master/src/timer.rs#L36 This needs to be hand-coded for the more specialized callbacks, but I don't think I need to implement most of those as they're built on top of the general callback framework. I may get rid of the `On[action]` traits and just have the callbacks be inherent methods, but I'm not sure.

You guys are using IUP to store the pointers but you spend some cycles constructing property strings whereas I just store the callbacks in a `HashMap` under the `*mut Ihandle` pointer. I do have some overhead in refcounting with the thread-local storage, I think I can avoid using TLS entirely but it requires some more unsafe code.

Additionally, I added a way to safely downcast an `Ihandle` to a more specific widget type that only exposes the functions valid for that type, and then the callback framework uses that to ensure the callback gets the exact widget type it applies to. I'm very proud of that part.

**Comment by thelink2012** [comment]
Ohh I see I didn't notice very well how your `callback_impl!` did things then now that you pointed it out.

Well I think that's all I had to ask, thanks, hope to see more ideias shared in the future :)

**Comment by ConnyOnny** [comment]
Sorry to revive this thread, but from a user's perspective: What should I use now? Are there tasks, where one or the other library is easier to use or preferable in another way or is the only difference between the two libraries alpha renaming (e.g. "children!" vs "elements!")?

**Comment by thelink2012** [comment]
I'm not sure either, the ideal would a pure rust toolkit, which @cybergeek94 seems to be working on as a different project I think.

**Comment by cybergeek94** [comment]
@ConnyOnny I meant to reply to your question a lot sooner. Right now I'm basically waiting for a couple different features to show up in Rust: specialization and something like what's being proposed in [this RFC](https://github.com/rust-lang/rfcs/pull/1546). Those two features would make designing an API like I want to provide in the direct-to-native KISS-UI 2.0 a bit easier.
